python/pytest-selfie/pytest_selfie/plugin.py

```diff
@@ -32,8 +32,23 @@
 
 
 class FSImplementation(FS):
-    def assert_failed(self, message: str, expected=None, actual=None) -> Exception:
-        raise Exception(message)
+    def assert_failed(self, message, expected=None, actual=None) -> Exception:
+        if expected is None and actual is None:
+            return AssertionError(message)
+
+        expected_str = self._nullable_to_string(expected, "(null)")
+        actual_str = self._nullable_to_string(actual, "(null)")
+
+        if not expected_str and not actual_str and (expected is None or actual is None):
+            return self._comparison_assertion(message, "(null)", "(null)")
+        return self._comparison_assertion(message, expected_str, actual_str)
+
+    def _nullable_to_string(self, value, on_null):
+        return str(value) if value is not None else on_null
+
+    def _comparison_assertion(self, message, expected, actual) -> Exception:
+        error_message = f"{message} - Expected: {expected}, Actual: {actual}"
+        return AssertionError(error_message)
```

Comment:
In pytest, a comparison assertion should be `assert expected == actual, msg`, see https://github.com/diffplug/selfie/issues/299#issue-2226257195